nifi-nar-bundles/nifi-druid-bundle/nifi-druid-processors/src/main/java/org/apache/nifi/processors/PutDruid.java

```diff
@@ -0,0 +1,206 @@
+						+ "Incoming flow files are expected to contain 1 or many JSON objects, one JSON object per line")
+public class PutDruid extends AbstractSessionFactoryProcessor {
+
+    private List<PropertyDescriptor> properties;
+    private Set<Relationship> relationships;
+    private final Map<Object,String> messageStatus = new HashMap<Object,String>();
+
+    public static final PropertyDescriptor DRUID_TRANQUILITY_SERVICE = new PropertyDescriptor.Builder()
+            .name("druid_tranquility_service")
+            .description("Tranquility Service to use for sending events to Druid")
+            .required(true)
+            .identifiesControllerService(DruidTranquilityService.class)
+            .build();
+
+    public static final Relationship REL_SUCCESS = new Relationship.Builder()
+            .name("SUCCESS")
+            .description("Succes relationship")
+            .build();
+
+    public static final Relationship REL_FAIL = new Relationship.Builder()
+            .name("FAIL")
+            .description("FlowFiles are routed to this relationship when they cannot be parsed")
+            .build();
+
+    public static final Relationship REL_DROPPED = new Relationship.Builder()
+            .name("DROPPED")
+            .description("FlowFiles are routed to this relationship when they are outside of the configured time window, timestamp format is invalid, ect...")
+            .build();
+
+    public void init(final ProcessorInitializationContext context){
+        List<PropertyDescriptor> properties = new ArrayList<>();
+        properties.add(DRUID_TRANQUILITY_SERVICE);
+        this.properties = Collections.unmodifiableList(properties);
+
+        Set<Relationship> relationships = new HashSet<Relationship>();
+        relationships.add(REL_SUCCESS);
+        relationships.add(REL_DROPPED);
+        relationships.add(REL_FAIL);
+        this.relationships = Collections.unmodifiableSet(relationships);
+    }
+
+    @Override
+    public Set<Relationship> getRelationships(){
+        return relationships;
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return properties;
+    }
+    
+    //Method breaks down incoming flow file and sends it to Druid Indexing servic
+    private void processFlowFile(ProcessContext context, ProcessSession session){
+    	 //Get handle on Druid Tranquility session
+    	 DruidTranquilityService tranquilityController = context.getProperty(DRUID_TRANQUILITY_SERVICE)
+                 .asControllerService(DruidTranquilityService.class);
+         Tranquilizer<Map<String,Object>> tranquilizer = tranquilityController.getTranquilizer();
+         
+         final FlowFile flowFile = session.get();
+         if (flowFile == null || flowFile.getSize() == 0) {
+             return;
+         }
+         
+         //Get data from flow file body
+         final byte[] buffer = new byte[(int) flowFile.getSize()];
```

Comment:
looks like we read in a bundle of messages and they're split on \r.  We really should refactor this to use the record reader paradigm which would greatly help the memory management issue and make this more portable for various incoming types of data.

Reply:
Any chance I can create a separate JIRA for that?